### ISSUE TITLE
`replace` command

### ISSUE THREAD

**Issue by pchiusano** [unisonweb/unison#318]
Example:

    > replace foo bar

This has the same effect as creating a new version of `foo` in a new file then doing `update` (which creates an edit for `foo`), followed by `> rename foo bar`. In other words, it lets you replace one definition with another, manually, without needing to do an `update`. This is another way of resolving a name conflict. If you just do a `rename foo#sdf bar`, dependents of the `foo#sdf` definition are unaffected. If you do a `replace`, you are marking `foo#sdf` as obsolete and saying you want all dependents to use `bar` instead.

This might also just be useful if you realize that a definition you already have lying around is what you want to use in place of something else, like if you realize you've written an overly concrete / monomorphic function that turns out to already be implemented in more generic form.

Notes:

* Either of the two arguments may be hash-qualified
* Should do a `propagate` afterwards

**Comment by hasheddan** [comment]
Hi @pchiusano! I recently discovered Unison and would love to start contributing. Would this be a good place to begin?

**Comment by aryairani** [comment]
Hi @hasheddan — yes and no :) We are in the middle of a refactor (`wip/branchless` branch), so this feature should get implemented there.  It would be in `HandleInput.hs` and you could likely use the hopefully correctly implemented `UpdateI` and `RenameTermI` as guidance.  

However, since the refactoring is super incomplete, super doesn’t run, and is  largely a mess, I’d hate to subject you to it.  Nevertheless, this feature is a ways down on our to-do list, so if you are feeling masochistic you could definitely take a stab at it without stepping on anyone’s toes.

We’ll try to answer any questions we can.

**Comment by runarorama** [comment]
Consider moving out of M1 to cut scope.

**Comment by aryairani** [comment]
@hasheddan This is ready to be worked on now, if you're still open to it. 😅

**Comment by pchiusano** [comment]
@aryairani how is this different than existing `replace.term` and `replace.type` commands?

**Comment by aryairani** [comment]
@pchiusano Good point, it took me a while to decide the answer yesterday before not closing it, but then I didn't write it down.

I think `replace.term` and `replace.type` just need to be updated to take any `HashQualified` (right now they only accept `ShortHash`, and we could also have a generic `replace` command that figures out whether the args are both a term or both a type, and calls the respective underlying logic.

**Comment by hasheddan** [comment]
@aryairani apologies for just circling back to this! Is this still something that needs to be worked on? If so, I would be happy to take it :) If not, I would be happy to work on anything else that is needing attention!

**Comment by runarorama** [comment]
I think this might be already mostly done, as `replace.term` and `replace.type` should now be able to take any `HashQualified`. The generic `replace` command might be nice to have though.

**Comment by tapegram** [comment]
@hasheddan are you currently working on this? I'd like to take a stab if not. Also, @aryairani @runarorama is this still relevant and useful?

**Comment by runarorama** [comment]
Yes, I'd say this is still relevant and useful. The current `replace.term` and `replace.type` commands could be unified under a single command `replace`. You'd have to check if the arguments are terms or types and give a helpful message if they're not both the same kind of thing.
